lib/amazonka/test/Main.hs

```diff
@@ -0,0 +1,97 @@
+import           Data.IORef                     (newIORef, atomicModifyIORef')
+import           Data.Time.Clock                (getCurrentTime, addUTCTime)
+import           Network.HTTP.Client            (HttpException(..))
+import           Network.HTTP.Types.Status      (status500)
+import           Test.Tasty
+import           Test.Tasty.HUnit
+
+-- Amazonka imports
+import           Amazonka                       (SessionToken)
+import           Amazonka.Auth.Background       (fetchAuthInBackground)
+import           Amazonka.Auth.Exception        (AuthError(..))
+import           Amazonka.Core                  (Time(..), ISO8601, Sensitive(..))
+import           Amazonka.Types                 (AuthEnv(..), ServiceError(..))
+
+main :: IO ()
+main = defaultMain tests
+
+tests :: TestTree
+tests = testGroup "Auth background error categorization"
+  [ testCase "UnknownAuthError wraps non-auth exceptions" $ do
+      menv <- mkAuthEnvAction (userError "random failure")
+      -- Note: fetchAuthInBackground immediately calls the provided action once
+      -- to seed the initial AuthEnv (this is the "first" call and succeeds),
+      -- then a background thread calls it again on refresh (the "subsequent"
+      -- call), which our helper makes fail with the supplied exception.
+      res <- try @AuthError $ do
+                _ <- fetchAuthInBackground menv
+                threadDelay 700000 -- give the background thread time to run (0.7s)
+      case res of
+        Left (UnknownAuthError (SomeException _)) -> pure ()
+        Left e -> assertFailure ("unexpected error: " ++ show e)
+        Right _ -> assertFailure "expected UnknownAuthError"
+  , testCase "AuthServiceError is propagated" $ do
+      let svc = ServiceError'
+                  { abbrev = "sts"
+                  , status = status500
+                  , headers = []
+                  , code = "InternalFailure"
+                  , message = Nothing
+                  , requestId = Nothing
+                  }
+      menv <- mkAuthEnvAction (AuthServiceError svc)
+      res <- try @AuthError $ do
+        _ <- fetchAuthInBackground menv
+        threadDelay 700000
+      case res of
+        Left (AuthServiceError s) -> case s of
+          ServiceError' { code = c } -> c @?= "InternalFailure"
+        Left e -> assertFailure ("unexpected error: " ++ show e)
+        Right _ -> assertFailure "expected AuthServiceError"
+  , testCase "RetrievalError is propagated" $ do
+      let retr = InvalidUrlException "url" "bad"
+      menv <- mkAuthEnvAction (RetrievalError retr)
+      res <- try @AuthError $ do
+                _ <- fetchAuthInBackground menv
+                threadDelay 700000
+      case res of
+        Left (RetrievalError _) -> pure ()
+        Left e -> assertFailure ("unexpected error: " ++ show e)
+        Right _ -> assertFailure "expected RetrievalError"
+  ]
+
+-- Helpers
```

Comment:
Please don't leave these marks in your code. It's distracting and they easily become out of date.